orderfilter/filter.go

```diff
@@ -0,0 +1,210 @@
+		return nil, err
+	}
+	rootOrderMessageSchema, err := messageLoader.Compile(rootOrderMessageSchemaLoader)
+	if err != nil {
+		return nil, err
+	}
+	return &Filter{
+		chainID:              chainID,
+		rawCustomOrderSchema: customOrderSchema,
+		orderSchema:          rootOrderSchema,
+		messageSchema:        rootOrderMessageSchema,
+	}, nil
+}
+
+func loadExchangeAddress(loader *jsonschema.SchemaLoader, chainID int) error {
+	contractAddresses, err := ethereum.GetContractAddressesForChainID(chainID)
+	if err != nil {
+		return err
+	}
+	// Note that exchangeAddressSchema accepts both checksummed and
+	// non-checksummed (i.e. all lowercase) addresses.
+	exchangeAddressSchema := fmt.Sprintf(`{"enum":[%q,%q]}`, contractAddresses.Exchange.Hex(), strings.ToLower(contractAddresses.Exchange.Hex()))
+	return loader.AddSchema("/exchangeAddress", jsonschema.NewStringLoader(exchangeAddressSchema))
+}
+
+func loadChainID(loader *jsonschema.SchemaLoader, chainID int) error {
+	chainIDSchema := fmt.Sprintf(`{"const":%d}`, chainID)
+	return loader.AddSchema("/chainId", jsonschema.NewStringLoader(chainIDSchema))
+}
+
+func newLoader(chainID int, customOrderSchema string) (*jsonschema.SchemaLoader, error) {
+	loader := jsonschema.NewSchemaLoader()
+	if err := loadChainID(loader, chainID); err != nil {
+		return nil, err
+	}
+	if err := loadExchangeAddress(loader, chainID); err != nil {
+		return nil, err
+	}
+	if err := loader.AddSchemas(builtInSchemas...); err != nil {
+		return nil, err
+	}
+	if err := loader.AddSchema("/customOrder", jsonschema.NewStringLoader(customOrderSchema)); err != nil {
+		return nil, err
+	}
+	return loader, nil
+}
+
+func NewFromTopic(topic string) (*Filter, error) {
+	// TODO(albrow): Use a cache for topic -> filter
+	var version int
+	var chainIDAndSchema string
+	if _, err := fmt.Sscanf(topic, topicVersionFormat, &version, &chainIDAndSchema); err != nil {
+		return nil, fmt.Errorf("could not parse topic version for topic: %q", topic)
+	}
+	if version != pubsubTopicVersion {
+		return nil, WrongTopicVersionError{
+			expectedVersion: pubsubTopicVersion,
+			actualVersion:   version,
+		}
+	}
+	var chainID int
+	var base64EncodedSchema string
+	if _, err := fmt.Sscanf(chainIDAndSchema, topicChainIDAndSchemaFormat, &chainID, &base64EncodedSchema); err != nil {
+		return nil, fmt.Errorf("could not parse chainID and schema from topic: %q", topic)
+	}
+	customOrderSchema, err := base64.URLEncoding.DecodeString(base64EncodedSchema)
+	if err != nil {
+		return nil, fmt.Errorf("could not base64-decode order schema: %q", base64EncodedSchema)
+	}
+	return New(chainID, string(customOrderSchema))
+}
+
+func (f *Filter) Topic() string {
+	if f.topic == "" {
+		f.topic = f.generateTopic()
+	}
+	return f.topic
+}
+
+func (v *Filter) generateTopic() string {
+	var holder interface{} = struct{}{}
+	_ = canonicaljson.Unmarshal([]byte(v.rawCustomOrderSchema), &holder)
+	canonicalOrderSchemaJSON, _ := canonicaljson.Marshal(holder)
```

Comment:
Can you add a comment to why this unmarshalling and marshalling is necessary?

Reply:
I added the following comment:

```go
	// Note(albrow): We use canonicaljson to elminate any differences in spacing,
	// formatting, and the order of field names. This ensures that two filters
	// that are semantically the same JSON object always encode to exactly the
	// same canonical topic string.
	//
	// So for example:
	//
	//     {
	//         "foo": "bar",
	//         "biz": "baz"
	//     }
	//
	// Will encode to the same topic string as:
	//
	//     {
	//         "biz":"baz",
	//         "foo":"bar"
	//     }
	//
```